topeft/modules/datacard_tools.py

```diff
@@ -306,6 +307,26 @@ def __init__(self,pkl_path,**kwargs):
         with open(topeft_path("params/wc_ranges.json"), "r") as wc_ranges_json:
             self.wc_ranges = json.load(wc_ranges_json)
 
+        with open(topeft_path("params/SMEFTsim-topU3l_dim6top.yml")) as f:
+            self.rotate = yaml.load(f,Loader=yaml.CLoader)
+            for k,v in self.rotate.items():
+                if isinstance(v, str):
+                    if "CW" in v or "SW" in v:
+                        v = eval(v.replace("{CW}", str(self.rotate["CW"])).replace("{SW}", str(self.rotate["SW"])))
+                        self.rotate[k] = v
+                elif isinstance(v, list):
+                    for i,vv in enumerate(v):
+                        if "CW" in vv or "SW" in vv:
+                            v[i] = vv.replace("{CW}", str(self.rotate["CW"])).replace("{SW}", str(self.rotate["SW"]))
+                    wcs = v[1]
+                    v   = v[0]
+                    wcs = {wc: wc+"[0, " + str(self.wc_ranges[wc][0]) + ", " + str(self.wc_ranges[wc][1]) + "]" for wc in wcs}
+                    v = v.format(**wcs)
+                    self.rotate[k] = v
+                if isinstance(v, str) and v in self.wc_ranges:
+                    v = v.replace(v, v+"[0, " + str(self.wc_ranges[v][0]) + ", " + str(self.wc_ranges[v][1]) + "]")
+                    self.rotate[k] = v
```

Comment:
This entire block does not need to be nested under the `with` block

Reply:
Good point

Reply:
The `if "CW" in v or "SW" in v` check is misleading, since the entire point of this block is to essentially evaluate the string `"np.sqrt(1 - np.square({CW}))"`. I guess it allows you to define the function for `CW` if you wanted to change it in the future, but that seems unnecessary to me?

Reply:
Yeah I was trying to keep it fancy but it wound up being a headache. We can probably just store the numerical values. I was thinking CW might change with other WCs, but it doesn't, it only gets multiplied by WCs (that we don't currently use).

Reply:
As per my continued war on python list comprehensions, I would propose to change this line:
```python
wcs = v[1]
v   = v[0]
wcs = {wc: wc+"[0, " + str(self.wc_ranges[wc][0]) + ", " + str(self.wc_ranges[wc][1]) + "]" for wc in wcs}
```
to the (imo) more readable
```python
wcs = {}
for wc in v[1]:
    lo,hi = self.wc_ranges[wc]
    wcs[wc] = f"{wc}[0,{lo},{hi}]"
v = v[0]
```

Reply:
Fair enough. I prefer comprehensions for speed, but this case is probably small enough that it doesn't matter.